### PR TITLE
🐛 Source Amazon Ads: Add missing reporting metrics to the source connector

### PR DESCRIPTION
**Pull request by Shishir-rmv**
## What

- The Amazon Ads source connector is missing some of the reporting metrics required to appropriately compute ad performance. This PR adds the necessary reporting metrics.

## How

- Added the missing reporting metrics for Sponsored Products, Sponsored Brands, Sponsored Display and Sponsored brands video as listed in the Amazon Ads [documentation](https://advertising.amazon.com/API/docs/en-us/reporting/v2/report-types)

## Recommended reading order
1. `airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/brands_report.py`
2. `airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/brands_video_report.py`
3. `airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/display_report.py`
4. `airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py`

## 🚨 User Impact 🚨
- No breaking changes

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

```
Results (4.89s):
      73 passed
Name                                                              Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------------
source_amazon_ads/__init__.py                                         2      0   100%
source_amazon_ads/constants.py                                        1      0   100%
source_amazon_ads/schemas/__init__.py                                 7      0   100%
source_amazon_ads/schemas/attribution_report.py                      21      0   100%
source_amazon_ads/schemas/common.py                                  51      1    98%   26
source_amazon_ads/schemas/profile.py                                 16      0   100%
source_amazon_ads/schemas/sponsored_brands.py                        22      0   100%
source_amazon_ads/schemas/sponsored_display.py                       31      0   100%
source_amazon_ads/schemas/sponsored_products.py                      37      0   100%
source_amazon_ads/source.py                                          44      1    98%   129
source_amazon_ads/streams/__init__.py                                 7      0   100%
source_amazon_ads/streams/attribution_report.py                      81      0   100%
source_amazon_ads/streams/common.py                                  79      1    99%   157
source_amazon_ads/streams/profiles.py                                21      0   100%
source_amazon_ads/streams/report_streams/__init__.py                  5      0   100%
source_amazon_ads/streams/report_streams/brands_report.py            10      0   100%
source_amazon_ads/streams/report_streams/brands_video_report.py      10      0   100%
source_amazon_ads/streams/report_streams/display_report.py           19      0   100%
source_amazon_ads/streams/report_streams/products_report.py          18      0   100%
source_amazon_ads/streams/report_streams/report_streams.py          240     19    92%   152, 205-206, 258-259, 378-379, 411-413, 418-419, 428-434
source_amazon_ads/streams/sponsored_brands.py                        26      0   100%
source_amazon_ads/streams/sponsored_display.py                       31      0   100%
source_amazon_ads/streams/sponsored_products.py                      41      0   100%
source_amazon_ads/utils.py                                           22      0   100%
unit_tests/__init__.py                                                0      0   100%
unit_tests/conftest.py                                               37      0   100%
unit_tests/test_attribution_report.py                                66      1    98%   47
unit_tests/test_report_streams.py                                   314      0   100%
unit_tests/test_source.py                                            47      0   100%
unit_tests/test_streams.py                                          126      1    99%   77
unit_tests/test_utils.py                                             11      0   100%
unit_tests/utils.py                                                  32      0   100%
-----------------------------------------------------------------------------------------------
TOTAL                                                              1475     24    98%
```

</details>

<details><summary><strong>Integration</strong></summary>

```
❯ python -m pytest integration_tests
Test session starts (platform: darwin, Python 3.9.6, pytest 6.2.5, pytest-sugar 0.9.6)
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/sverma/code/github.com/Shishir-rmv/airbyte/airbyte-integrations/connectors/source-amazon-ads/.hypothesis/examples')
rootdir: /Users/sverma/code/github.com/Shishir-rmv/airbyte, configfile: pytest.ini
plugins: hypothesis-6.54.6, requests-mock-1.9.3, mock-3.7.0, sugar-0.9.6, timeout-1.4.2, cov-3.0.0
collecting ...

Results (0.14s):
```

</details>

<details><summary><strong>Acceptance</strong></summary>

```
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_airbyte_trace_message_on_failure[inputs1] ✓                                                                                     85% ████████▌ {"type": "LOG", "log": {"level": "INFO", "message": "The configured catalog is built from a custom configured catalog."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_full_refresh.py::TestFullRefresh.test_sequential_reads[inputs0] ✓                                                                                           88% ████████▉ {"type": "LOG", "log": {"level": "INFO", "message": "The configured catalog is built from a custom configured catalog."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_full_refresh.py::TestFullRefresh.test_sequential_reads[inputs1] ✓                                                                                           91% █████████▎{"type": "LOG", "log": {"level": "INFO", "message": "The configured catalog is built from a custom configured catalog."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_two_sequential_reads[inputs0] ✓                                                                                        94% █████████▌{"type": "LOG", "log": {"level": "INFO", "message": "The configured catalog is built from a custom configured catalog."}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_read_sequential_slices[inputs0] ✓                                                                                      97% █████████▊{"type": "LOG", "log": {"level": "INFO", "message": "The configured catalog is built from a custom configured catalog."}}


―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― TestIncremental.test_state_with_abnormally_large_values[inputs0] ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

self = <source_acceptance_test.tests.test_incremental.TestIncremental object at 0x107c319d0>, connector_config = SecretDict(******)
configured_catalog = ConfiguredAirbyteCatalog(streams=[ConfiguredAirbyteStream(stream=AirbyteStream(name='sponsored_products_report_stream'...cremental'>, cursor_field=None, destination_sync_mode=<DestinationSyncMode.overwrite: 'overwrite'>, primary_key=None)])
future_state = [{'stream': {'stream_descriptor': {'name': 'sponsored_products_report_stream'}, 'stream_state': {'1861552880916640': {'reportDate': '20990101'}}}, 'type': 'STREAM'}]
docker_runner = <source_acceptance_test.utils.connector_runner.ConnectorRunner object at 0x10a0a2820>

    def test_state_with_abnormally_large_values(self, connector_config, configured_catalog, future_state, docker_runner: ConnectorRunner):
        configured_catalog = incremental_only_catalog(configured_catalog)
        output = docker_runner.call_read_with_state(config=connector_config, catalog=configured_catalog, state=future_state)
        records = filter_output(output, type_=Type.RECORD)
        states = filter_output(output, type_=Type.STATE)

>       assert (
            not records
        ), f"The sync should produce no records when run with the state with abnormally large values {records[0].record.stream}"
E       AssertionError: The sync should produce no records when run with the state with abnormally large values sponsored_products_report_stream
E       assert not [AirbyteMessage(type=<Type.RECORD: 'RECORD'>, log=None, spec=None, connectionStatus=None, catalog=None, record=Airbyte...: None, 'attributedConversions1dSameSKU': '0'}}, emitted_at=1669095063456), state=None, trace=None, control=None), ...]

../../bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py:256: AssertionError
--------------------------------------------------------------------------------------------------------------------- Captured log setup ----------------------------------------------------------------------------------------------------------------------
INFO     root:common.py:175 The configured catalog is built from a custom configured catalog.

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental.test_state_with_abnormally_large_values[inputs0] ⨯                                                                         100% ██████████
{"type": "LOG", "log": {"level": "INFO", "message": "/Users/sverma/code/github.com/Shishir-rmv/airbyte/airbyte-integrations/connectors/source-amazon-ads - SAT run - 386cdcf937310a56440290a5bd69d6e23181ac84 - FAILED"}}

====================================================================================================================== warnings summary =======================================================================================================================
../../bases/source-acceptance-test/source_acceptance_test/config.py:254: 30 warnings
airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py: 1 warning
  /Users/sverma/code/github.com/Shishir-rmv/airbyte/airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/config.py:254: DeprecationWarning: The 'warn' function is deprecated, use 'warning' instead
    logging.warn("The acceptance-test-config.yml file is in a legacy format. Please migrate to the latest format.")

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=================================================================================================================== short test summary info ===================================================================================================================
FAILED ../../bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead::test_read[inputs0] - Failed: Stream sponsored_display_ad_groups: All expected records must be produced
FAILED ../../bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py::TestIncremental::test_state_with_abnormally_large_values[inputs0] - AssertionError: The sync should produce no records when run with the state with abnormally l...

Results (498.67s):
      32 passed
       2 failed
         - airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:644 TestBasicRead.test_read[inputs0]
         - airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py:250 TestIncremental.test_state_with_abnormally_large_values[inputs0]
```

</details>

### CODE REVIEW

**Comment by Shishir-rmv** [comment]
This solves - https://github.com/airbytehq/airbyte/issues/18025

**Comment by marcosmarxm** [comment]
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

File: airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/brands_report.py
Comment (monai) [inline airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/brands_report.py:47]: I see that you added the missing metrics that are defined in the documentation https://advertising.amazon.com/API/docs/en-us/reporting/v2/report-types. But for some record types, the current code has redundant metrics. For example, in this list, there are `targetId`, `targetingExpression`, `targetingText`, `targetingType`. However, the documentation doesn't indicate they are available for this particular record type. Does the report return any values for them? If not, such metrics that aren't listed in the documentation should be removed.

Please check all the report types, all record types, and all the metrics. The code should exactly reflect the documentation except when the documentation is wrong, then follow the actual behavior of the reporting system.

File: airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py
Comment (monai) [inline airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py:285]: Add a comment on why this `if` statement is needed. Explain what happens if it's missing and what happens when `profile.accountInfo.type` is other than `seller`. In addition, add a link to relevant documentation on this behavior if available.

**Review by monai — CHANGES_REQUESTED** [review]
Thank you for your contribution. I want to ask you to make some changes before it can be merged.

Add `test_strictness_level: high` to `acceptance-test-config.yml` to resolve a [failing check](https://github.com/airbytehq/airbyte/actions/runs/3563527564/jobs/5986411219). See the [documentation](https://docs.airbyte.com/connector-development/testing-connectors/source-acceptance-tests-reference/#strictness-level) for more info.

File: airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/brands_report.py
Comment (Shishir-rmv) [reply airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/brands_report.py:47]: @monai - Thanks for your feedback. I am working on it and will update the PR soon.

Comment (Shishir-rmv) [reply airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/brands_report.py:43]: @monai - I updated the PR and deleted the metrics that are no longer available or supported. There are also certain fields indicated in the documentation but not supported, such as 'attributedKindleEditionNormalizedPagesRoyalties14d' and 'attributedKindleEditionNormalizedPagesRead14d' for Sponsored products, which I have not included in the metrics.

Please review. Thank you.

File: airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py
Comment (Shishir-rmv) [reply airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py]: @monai - Added comment

File: airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py
Comment (monai) [inline airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py:285]:
```suggestion
            metrics_list = [m for m in metrics_list if m != "sku"]
```
List comprehension would be more pythonic since it constructs a new list in one pass and doesn't mutate the copy of the list

File: airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py
Comment (monai) [inline airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py:286]:
```suggestion
```

File: airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/display_report.py
Comment (monai) [inline airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/display_report.py:238]: See my comment for `products_report.py`

File: airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py
Comment (monai) [inline airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py:283]: Add a unit test for this `if` and assert the removal of `sku` metric. And the same with `display_report.py`.

**Review by monai — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py
Comment (Shishir-rmv) [reply airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py]: @monai Thanks for the comment. I have updated the code.

File: airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py
Comment (Shishir-rmv) [reply airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/products_report.py:280]: @monai - I've added the unit test to assert the presence of `sku` metric in the report requests based on profiles.

**Comment by Shishir-rmv** [comment]
@monai - requesting to kindly review this again. Thank you.

**Review by monai — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-amazon-ads/unit_tests/test_report_streams.py
Comment (Shishir-rmv) [reply airbyte-integrations/connectors/source-amazon-ads/unit_tests/test_report_streams.py]: @monai - Thank you for the feedback. I have fixed the test cases as you requested, please review.

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

File: airbyte-integrations/connectors/source-amazon-ads/unit_tests/test_report_streams.py
Comment (monai) [inline airbyte-integrations/connectors/source-amazon-ads/unit_tests/test_report_streams.py:187]:
```suggestion
            if record_type == RecordType.PRODUCTADS or record_type == RecordType.ASINS:
```

Looks good overall. It seems like this is the last of my change request.

After it, please run `./gradlew format` in the Airbyte root directory to fix all formatting issues and commit all the changes it will make in the `source-amazon-ads` directory. I see some formatting issues in your code, for example, a single comma. This tool will fix them.

**Review by monai — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-amazon-ads/unit_tests/test_report_streams.py
Comment (Shishir-rmv) [reply airbyte-integrations/connectors/source-amazon-ads/unit_tests/test_report_streams.py]: @monai - Thank you for your feedback. I have made the changes. Please review.

**Review by monai — CHANGES_REQUESTED** [review]

**Review by monai — APPROVED** [review]

**Comment by monai** [comment]
Good job! Since this connector's level of maturity is [General Availability](https://docs.airbyte.com/project-overview/product-release-stages/#general-availability-ga), it requires a second review from the Airbyte team before merging and publishing. Yet, if you want to use it now, you can always build the Docker image from your branch, push it to your repository and add it as a custom connector.

cc @sajarin

**Comment by Shishir-rmv** [comment]
@grubberr - requesting to kindly review this PR. Thank you.

**Comment by Shishir-rmv** [comment]
@grubberr / @monai / @sajarin - Any update on this? There appear to be conflicts presently because other PRs were merged before this. The major version has been upgraded to 1.0.0. Should I bump this up to 1.0.1?

**Comment by Shishir-rmv** [comment]
@sherifnada - requesting to kindly review this PR. Thank you.

**Comment by sajarin** [comment]
Hey thanks for the contribution and apologies for the delay! 

I'll be the DRI responsible for reviewing this and getting it across the finish line. We currently have a backlog of about a dozen PRs, but just wanted to let you know that this PR has been added to the queue and I'll be trying my best to get to this as soon as possible. 

Thanks for the contribution and for being patient :)

**Comment by Shishir-rmv** [comment]
> @Shishir-rmv I'm closing this because is open for a long time. The connector now uses `v3` from Amazon Ads API and there are a lot of changes. Let me know if you want to continue this work in a new branch if so I'll prioritize to review and merge ASAP.

@marcosmarxm - I got this change on a new branch here: https://github.com/airbytehq/airbyte/pull/25412. Please prioritize. Thanks.

**Pull request closed without merging**
